internal/state/listener.go

```diff
@@ -69,114 +78,213 @@ func newHTTPSListenerConfigurator(
 }
 
 func (c *httpsListenerConfigurator) configure(gl v1beta1.Listener) *listener {
-	var path string
-	var err error
+	validate := func(gl v1beta1.Listener) []conditions.Condition {
+		return validateHTTPSListener(gl, c.gateway.Namespace)
+	}
 
-	valid := validateHTTPSListener(gl, c.gateway.Namespace)
+	l := configureListenerAndUseHostname(gl, c.gateway, c.usedHostnames, validate)
 
-	if valid {
-		nsname := types.NamespacedName{
-			Namespace: c.gateway.Namespace,
-			Name:      string(gl.TLS.CertificateRefs[0].Name),
-		}
+	if !l.Valid {
+		return l
+	}
 
-		path, err = c.secretMemoryMgr.Request(nsname)
-		if err != nil {
-			valid = false
-		}
+	nsname := types.NamespacedName{
+		Namespace: c.gateway.Namespace,
+		Name:      string(gl.TLS.CertificateRefs[0].Name),
 	}
 
-	h := getHostname(gl.Hostname)
+	path, err := c.secretMemoryMgr.Request(nsname)
+	if err != nil {
+		l.Valid = false
 
-	if holder, exist := c.usedHostnames[h]; exist {
-		valid = false
-		holder.Valid = false // all listeners for the same hostname become conflicted
-	}
+		msg := fmt.Sprintf("Failed to get the certificate %s: %v", nsname.String(), err)
+		l.Conditions = append(l.Conditions, conditions.NewListenerInvalidCertificateRef(msg)...)
 
-	l := &listener{
-		Source:            gl,
-		Valid:             valid,
-		SecretPath:        path,
-		Routes:            make(map[types.NamespacedName]*route),
-		AcceptedHostnames: make(map[string]struct{}),
+		return l
 	}
 
-	c.usedHostnames[h] = l
+	l.SecretPath = path
 
 	return l
 }
 
 type httpListenerConfigurator struct {
 	usedHostnames map[string]*listener
+	gateway       *v1beta1.Gateway
 }
 
-func newHTTPListenerConfigurator() *httpListenerConfigurator {
+func newHTTPListenerConfigurator(gw *v1beta1.Gateway) *httpListenerConfigurator {
 	return &httpListenerConfigurator{
 		usedHostnames: make(map[string]*listener),
+		gateway:       gw,
 	}
 }
 
 func (c *httpListenerConfigurator) configure(gl v1beta1.Listener) *listener {
-	valid := validateHTTPListener(gl)
+	return configureListenerAndUseHostname(gl, c.gateway, c.usedHostnames, validateHTTPListener)
+}
 
-	h := getHostname(gl.Hostname)
+type invalidProtocolListenerConfigurator struct{}
 
-	if holder, exist := c.usedHostnames[h]; exist {
+func newInvalidProtocolListenerConfigurator() *invalidProtocolListenerConfigurator {
+	return &invalidProtocolListenerConfigurator{}
+}
+
+func (c *invalidProtocolListenerConfigurator) configure(gl v1beta1.Listener) *listener {
+	msg := fmt.Sprintf("Protocol %q is not supported, use %q or %q",
+		gl.Protocol, v1beta1.HTTPProtocolType, v1beta1.HTTPSProtocolType)
+
+	return &listener{
+		Source:            gl,
+		Valid:             false,
+		Routes:            make(map[types.NamespacedName]*route),
+		AcceptedHostnames: make(map[string]struct{}),
+		Conditions: []conditions.Condition{
+			conditions.NewListenerUnsupportedProtocol(msg),
+		},
+	}
+}
+
+func configureListenerAndUseHostname(
```

Comment:
merging of two configurators left a singe `configure` method that includes that code now. I don't think it is too terrible now - it invokes the protocol-specific listener validations, runs cross-listener validation, requests certs for TLS listener. ultimately, it configures a listener.

Reply:
> Can we split this function up so that we handle all the validation in a single function, and then in a separate function configure the listener?

I split it into 3 functions. hopefully it is better now